### ISSUE TITLE
Allow `$_` as an unused variable identifier

### ISSUE THREAD

**Issue by JulienPalard** [phpmd/phpmd#326]
Thinking about https://github.com/phpmd/phpmd/issues/48, and `allow-unused-foreach-variables` why not allowing `$_` (or typically a configurable whitelist set) as a common name for "unused variable", a bit like in `erlang`, allowing:

    foreach ($itemIds as $id => $_)

and

    list($foo, $_, $baz) = $this->getFooBarBaz();

to be written without PHPMD complaining ?

The underlying question is, does one prefer:

    foreach ($itemsIds as $id => $_)

over

    foreach (array_keys($itemsIds) as $id)

I prefer the first, other probably prefer the last, why not dropping this as an option ?

**Comment by ravage84** [comment]
@JulienPalard I don't think this is a good solution.
a) It's not very readable
b) It's not self explaining
c) You would violate against the ShortVaraible rule
d) It's an implicit behavior, only made for this particular case in a particular tool (PHPMD)

**Comment by JulienPalard** [comment]
So, do you really prefer calling `array_keys()` each time ?

a) I don't think array_keys is more readable: It's longer, it's a set of parenthesis for nothing.
b) It's not self-explaining for the developper ignoring the practice, but it's a commonly used practice in loads of languages, so if a PHP developper still don't know it, it's a good time to introduce it via phpmd so he will know it when using another language:
 - Python: [Not sure if this is a Eclipse thing or not, but I generally use '_' to denote values I don't care ](http://stackoverflow.com/questions/11486148/unused-variable-naming-in-python)
 - Python: [_ is a standard placeholder name for ignored members](http://stackoverflow.com/questions/5477134/how-can-i-get-around-declaring-an-unused-variable-in-a-for-loop)
 - Python: [Using the underscore for unused variables is definitely acceptable.](http://programmers.stackexchange.com/questions/139582/which-style-to-use-for-unused-return-parameters-in-a-python-function-call)
 - Ruby [Use _ for unused block parameters](https://github.com/bbatsov/ruby-style-guide/issues/237)
 - Ruby [Now _ is the designated variable name to use when you don't want to use a value. ](http://stackoverflow.com/questions/8689673/marking-an-unused-block-variable)
 - Ruby [So best practice is to use "_" for all unused variables](https://bugs.ruby-lang.org/issues/6693)
 - Erlang (From the documentation itself) [The anonymous variable is denoted by underscore (_) and can be used when a variable is required but its value can be ignored.](http://erlang.org/doc/reference_manual/expressions.html)
 - JS [Single underscore looks perfectly fine as a throw-away argument.](https://github.com/jshint/jshint/issues/1762)
 - Rust From the doc itself: [these warnings can
    // be silenced by prefixing the variable name with an underscore](http://rustbyexample.com/variable_bindings.html)
 - Haskell (From the compiler itself) [It is a convention that the compiler will not warn about unused variables that begin with an underscore](https://prime.haskell.org/wiki/Underscore)

c) It's easy to add an exception to this rule, like `$x`, `$y`, `$i`, etc...

d) As seen in `b)` it's not a particular case in a particular tool, it's a commonly accepted and used practice.

**Comment by ravage84** [comment]
Using ``array_keys()`` seems to be a valid way, as it explicitly explains what the code does.
But it's not "nice" looking, I agree.

It's good if programmers of other languages have settled on this matter in their language but in PHP "we" havn't , yet.
And since using a leading underscore can mean protected/private in case of properties, it still would not be a good choice, IMHO.

Anyway, I guess adding an ``exceptions`` parameter which defaults to empty to the UnusedLocalVariable rule is what will solve your issue and still is more flexible than enforcing an agreed upon standard from othe rlanguages.

I'd be happy to accept a PR which implements this new config parameter.

**Comment by zra95** [comment]
In swift: _ mean unused too ;)

**Comment by amirouche** [comment]
In Scheme: _ means anything or throw away value.

**Comment by ravage84** [comment]
Guys, I don't need any further example, once https://github.com/phpmd/phpmd/pull/329 is merged, you can configure your own exceptions/dummy variable names.

**Comment by JulienBreux** [comment]
In GoLang: _ means throw away value.

**Comment by still-dreaming-1** [comment]
@ravage84 I'm confused, how we actually apply this option? The documentation shows the rule and the option, but does not provide any examples of applying it. I tried adding it to my xml file, but it didn't work, so I guess I did it wrong because I was just guessing how to do it.

I also still agree with the original request of making unused $_ ignored by default. It is not just something from other languages, it is a convention in PHP as well. Here is an [example](https://github.com/vimeo/psalm/blob/5471b89391f93712d9a13cfce354a02b5ebf7854/src/Psalm/Checker/StatementsChecker.php#L543) of a PHP code quality tool that ignores unused $_ variables.

**Comment by kylekatarnls** [comment]
You should have in your phpmd.xml:
```xml
<rule ref="rulesets/unusedcode.xml" />
```
You can replace this with:
```xml
<rule ref="rulesets/unusedcode.xml">
    <exclude name="UnusedLocalVariable" />
</rule>
<rule ref="rulesets/unusedcode.xml/UnusedLocalVariable">
    <properties>
        <property name="exceptions" value="_" />
    </properties>
</rule>
```
So `$_` no longer raise `UnusedLocalVariable` violation.

**Comment by paulshryock** [comment]
Need this for `UnusedFormalParameter` as well.
